File: loopgraph_supervisor/loopspec.py
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

NodeKind = Literal["dsh_execute", "verifier", "human_gate", "promotion", "terminal"]
Outcome = Literal["pass", "fail", "retry", "approve", "auto_promote", "reject", "exhausted"]


@dataclass(frozen=True)
class LoopNode:
    id: str
    kind: NodeKind
    role: str | None = None


@dataclass(frozen=True)
class LoopEdge:
    source: str
    target: str
    outcomes: tuple[Outcome, ...] = ()


@dataclass(frozen=True)
class LoopSpec:
    spec_id: str
    revision: int
    entrypoint: str
    nodes: tuple[LoopNode, ...]
    edges: tuple[LoopEdge, ...]
    max_iterations: int
    predecessor_hash: str | None = None

    def __post_init__(self) -> None:
        if not self.spec_id or not isinstance(self.revision, int) or isinstance(self.revision, bool) or self.revision < 1 or not isinstance(self.max_iterations, int) or isinstance(self.max_iterations, bool) or self.max_iterations < 1:
            raise ValueError("LoopSpec requires a non-empty id, positive revision, and max_iterations")
        valid_kinds = {"dsh_execute", "verifier", "human_gate", "promotion", "terminal"}
        valid_outcomes = {"pass", "fail", "retry", "approve", "auto_promote", "reject", "exhausted"}
        if any(node.kind not in valid_kinds for node in self.nodes) or any(outcome not in valid_outcomes for edge in self.edges for outcome in edge.outcomes):
            raise ValueError("LoopSpec contains an unsupported node kind or outcome")
        node_ids = [node.id for node in self.nodes]
        if len(node_ids) != len(set(node_ids)):
            raise ValueError("LoopSpec node ids must be unique")
        if self.entrypoint not in set(node_ids):
            raise ValueError(f"LoopSpec entrypoint is not defined: {self.entrypoint}")
        for edge in self.edges:
            if edge.source not in set(node_ids) or edge.target not in set(node_ids):
                raise ValueError(f"LoopSpec edge references an unknown node: {edge.source}->{edge.target}")
            if len(edge.outcomes) != len(set(edge.outcomes)):
                raise ValueError(f"LoopSpec edge outcomes must be unique: {edge.source}->{edge.target}")
        routes = [(edge.source, outcome) for edge in self.edges for outcome in edge.outcomes]
        if len(routes) != len(set(routes)):
            raise ValueError("LoopSpec source/outcome routes must be unique")

    def next_node(self, source: str, outcome: Outcome) -> str:
        matches = [edge.target for edge in self.edges if edge.source == source and outcome in edge.outcomes]
        if len(matches) != 1:
            raise ValueError(f"LoopSpec has no unique edge for {source} on {outcome}")
        return matches[0]

    def document(self) -> dict[str, Any]:
        return {
            "schema_version": 1,
            "spec_id": self.spec_id,
            "revision": self.revision,
            "predecessor_hash": self.predecessor_hash,
            "entrypoint": self.entrypoint,
            "max_iterations": self.max_iterations,
            "nodes": [{key: value for key, value in {"id": node.id, "kind": node.kind, "role": node.role}.items() if value is not None} for node in self.nodes],
            "edges": [{"source": edge.source, "target": edge.target, "outcomes": list(edge.outcomes)} for edge in self.edges],
        }

    def content_hash(self) -> str:
        encoded = json.dumps(self.document(), ensure_ascii=False, sort_keys=True, separators=(",", ":"))
        return hashlib.sha256(encoded.encode()).hexdigest()


def load_loopspec(path: str | Path) -> LoopSpec:
    """Load one versioned LoopGraph artifact and validate it before use."""
    document = json.loads(Path(path).read_text())
    if not isinstance(document, dict):
        raise ValueError("LoopSpec artifact must contain one JSON object")
    if set(document) != {"schema_version", "spec_id", "revision", "predecessor_hash", "entrypoint", "max_iterations", "nodes", "edges"}:
        raise ValueError("LoopSpec artifact has an unexpected shape")
    if document.get("schema_version") != 1:
        raise ValueError("unsupported LoopSpec schema version")
    if any(set(item) - {"id", "kind", "role"} for item in document["nodes"]) or any(set(item) - {"source", "target", "outcomes"} for item in document["edges"]):
        raise ValueError("LoopSpec nodes or edges contain unknown fields")
    spec = LoopSpec(
        spec_id=document["spec_id"],
        revision=document["revision"],
        predecessor_hash=document.get("predecessor_hash"),
        entrypoint=document["entrypoint"],
        max_iterations=document["max_iterations"],
        nodes=tuple(LoopNode(item["id"], item["kind"], item.get("role")) for item in document["nodes"]),
        edges=tuple(LoopEdge(item["source"], item["target"], tuple(item.get("outcomes", []))) for item in document["edges"]),
    )
    from .graph_validator import validate_loopgraph

    validate_loopgraph(spec, require_coding_supervisor=spec.spec_id == "coding-supervisor")
    return spec


def load_active_loopspec(path: str | Path) -> LoopSpec:
    manifest_path = Path(path).resolve()
    document = json.loads(manifest_path.read_text())
    if not isinstance(document, dict) or set(document) != {"schema_version", "spec_id", "active_revision", "artifact", "content_hash"} or document.get("schema_version") != 1:
        raise ValueError("active LoopSpec manifest has an unexpected shape")
    artifact_name = document["artifact"]
    if not isinstance(artifact_name, str) or Path(artifact_name).name != artifact_name or not artifact_name.endswith(".json"):
        raise ValueError("active LoopSpec artifact must be a safe sibling JSON file")
    artifact_path = (manifest_path.parent / artifact_name).resolve()
    try:
        artifact_path.relative_to(manifest_path.parent)
    except ValueError as error:
        raise ValueError("active LoopSpec artifact escapes its manifest directory") from error
    spec = load_loopspec(artifact_path)
    if spec.spec_id != document["spec_id"] or spec.revision != document["active_revision"] or spec.content_hash() != document["content_hash"]:
        raise ValueError("active LoopSpec manifest does not bind its artifact")
    return spec


def coding_spec_revision(revision: int) -> LoopSpec:
    artifact = Path(__file__).resolve().parents[1] / "configs" / "loopspecs" / "coding-supervisor" / f"v{revision}.json"
    return load_loopspec(artifact)


def coding_spec_chain() -> tuple[LoopSpec, ...]:
    active = default_coding_spec()
    specs = tuple(coding_spec_revision(revision) for revision in range(1, active.revision + 1))
    for predecessor, candidate in zip(specs, specs[1:]):
        if candidate.predecessor_hash != predecessor.content_hash():
            raise ValueError("repository LoopSpec predecessor chain is invalid")
    return specs


def default_coding_spec() -> LoopSpec:
    manifest = Path(__file__).resolve().parents[1] / "configs" / "loopspecs" / "coding-supervisor" / "active.json"
    return load_active_loopspec(manifest)

File: packages/dsh-loopgraph-supervisor/lib/loopspec.js
import { createHash } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
function canonical(value) {
    if (Array.isArray(value))
        return `[${value.map(canonical).join(',')}]`;
    if (value && typeof value === 'object')
        return `{${Object.entries(value).sort(([a], [b]) => a.localeCompare(b)).map(([key, item]) => `${JSON.stringify(key)}:${canonical(item)}`).join(',')}}`;
    return JSON.stringify(value);
}
export function loopSpecHash(spec) {
    return createHash('sha256').update(canonical(spec)).digest('hex');
}
export function loadLoopSpec(path = fileURLToPath(new URL('../loopspec.v1.json', import.meta.url))) {
    const raw = JSON.parse(readFileSync(path, 'utf8'));
    const allowed = new Set(['schema_version', 'spec_id', 'revision', 'predecessor_hash', 'entrypoint', 'max_iterations', 'nodes', 'edges']);
    if (Object.keys(raw).some(key => !allowed.has(key)))
        throw new Error('LoopSpec contains unknown fields');
    const value = {
        schema_version: raw.schema_version,
        spec_id: raw.spec_id,
        revision: raw.revision,
        predecessor_hash: raw.predecessor_hash,
        entrypoint: raw.entrypoint,
        max_iterations: raw.max_iterations,
        nodes: raw.nodes,
        edges: raw.edges,
    };
    if (value.nodes.some(node => Object.keys(node).some(key => !['id', 'kind', 'role'].includes(key))) || value.edges.some(edge => Object.keys(edge).some(key => !['source', 'target', 'outcomes'].includes(key))))
        throw new Error('LoopSpec nodes or edges contain unknown fields');
    validateLoopSpec(value);
    return value;
}
function activePath(workspace) {
    const root = process.env.DSH_HOME || join(homedir(), '.dsh');
    const key = createHash('sha256').update(workspace).digest('hex');
    return join(root, 'loopgraph', 'active-specs', `${key}.json`);
}
export function loadWorkspaceLoopSpec(workspace, fallback) {
    const path = activePath(workspace);
    return existsSync(path) ? loadLoopSpec(path) : fallback;
}
export function saveWorkspaceLoopSpec(workspace, spec) {
    validateLoopSpec(spec);
    const path = activePath(workspace);
    mkdirSync(dirname(path), { recursive: true, mode: 0o700 });
    const temporary = `${path}.${process.pid}.tmp`;
    writeFileSync(temporary, `${JSON.stringify(spec, null, 2)}\n`, { mode: 0o600 });
    renameSync(temporary, path);
}
export function nodeForRole(spec, role) {
    const matches = spec.nodes.filter(node => node.role === role || (!node.role && node.id === role));
    if (matches.length !== 1)
        throw new Error(`LoopSpec has no unique runtime role: ${role}`);
    return matches[0].id;
}
export function nextNode(spec, source, outcome, iteration) {
    if (iteration < 0)
        throw new Error('LoopSpec iteration cannot be negative');
    if (iteration >= spec.max_iterations && ['fail', 'retry'].includes(outcome))
        throw new Error(`LoopSpec iteration limit exceeded: ${spec.max_iterations}`);
    const matches = spec.edges.filter(edge => edge.source === source && edge.outcomes.includes(outcome));
    if (matches.length !== 1)
        throw new Error(`LoopSpec has no unique edge for ${source} on ${outcome}`);
    return matches[0].target;
}
export function validateLoopSpec(spec) {
    if (spec.schema_version !== 1 || !spec.spec_id || !Number.isSafeInteger(spec.revision) || spec.revision < 1 || !Number.isSafeInteger(spec.max_iterations) || spec.max_iterations < 1)
        throw new Error('invalid LoopSpec identity or limits');
    const ids = spec.nodes.map(node => node.id);
    if (new Set(ids).size !== ids.length || !ids.includes(spec.entrypoint))
        throw new Error('invalid LoopSpec nodes or entrypoint');
    const kinds = new Set(['dsh_execute', 'verifier', 'human_gate', 'promotion', 'terminal']);
    const validOutcomes = new Set(['pass', 'fail', 'retry', 'approve', 'auto_promote', 'reject', 'exhausted']);
    const requiredOutcomes = { dsh_execute: ['pass', 'fail', 'exhausted'], verifier: ['fail', 'retry', 'approve', 'auto_promote', 'exhausted'], human_gate: ['approve', 'retry', 'reject'], promotion: ['pass'] };
    const routes = new Set();
    const outgoing = new Map();
    const nodeOutcomes = new Map();
    for (const node of spec.nodes) {
        if (!kinds.has(node.kind))
            throw new Error(`unsupported LoopSpec node kind: ${node.kind}`);
        outgoing.set(node.id, new Set());
        nodeOutcomes.set(node.id, new Set());
    }
    for (const edge of spec.edges) {
        if (!ids.includes(edge.source) || !ids.includes(edge.target) || edge.outcomes.length === 0)
            throw new Error('LoopSpec edge references unknown nodes or no outcomes');
        outgoing.get(edge.source)?.add(edge.target);
        for (const outcome of edge.outcomes) {
            if (!validOutcomes.has(outcome))
                throw new Error(`unsupported LoopSpec outcome: ${outcome}`);
            const key = `${edge.source}\0${outcome}`;
            if (routes.has(key))
                throw new Error(`duplicate LoopSpec route: ${edge.source}/${outcome}`);
            routes.add(key);
            nodeOutcomes.get(edge.source)?.add(outcome);
        }
    }
    const reachable = new Set([spec.entrypoint]);
    const queue = [spec.entrypoint];
    while (queue.length)
        for (const target of outgoing.get(queue.shift()) ?? [])
            if (!reachable.has(target)) {
                reachable.add(target);
                queue.push(target);
            }
    if (reachable.size !== ids.length)
        throw new Error('LoopSpec contains unreachable nodes');
    const terminals = spec.nodes.filter(node => node.kind === 'terminal').map(node => node.id);
    if (!terminals.length)
        throw new Error('LoopSpec requires a terminal node');
    for (const node of spec.nodes) {
        const count = outgoing.get(node.id)?.size ?? 0;
        if (node.kind === 'terminal' && count)
            throw new Error(`terminal node has outgoing edges: ${node.id}`);
        if (node.kind !== 'terminal' && !count)
            throw new Error(`non-terminal node has no outgoing edge: ${node.id}`);
        for (const outcome of requiredOutcomes[node.kind] ?? [])
            if (!nodeOutcomes.get(node.id)?.has(outcome))
                throw new Error(`node ${node.id} is missing required outcome: ${outcome}`);
    }
    const canTerminate = new Set(terminals);
    let changed = true;
    while (changed) {
        changed = false;
        for (const [source, targets] of outgoing)
            if (!canTerminate.has(source) && [...targets].some(target => canTerminate.has(target))) {
                canTerminate.add(source);
                changed = true;
            }
    }
    if (canTerminate.size !== ids.length)
        throw new Error('LoopSpec contains nodes with no terminating path');
    if (spec.entrypoint !== nodeForRole(spec, 'execute'))
        throw new Error('LoopSpec entrypoint must have the execute role');
    const roleKinds = { execute: 'dsh_execute', verify: 'verifier', human_gate: 'human_gate', promote: 'promotion', complete: 'terminal', failed: 'terminal' };
    for (const [role, kind] of Object.entries(roleKinds)) {
        const matches = spec.nodes.filter(node => node.role === role);
        if (matches.length !== 1 || matches[0].kind !== kind)
            throw new Error(`LoopSpec role ${role} requires one ${kind} node`);
    }
    const roles = Object.fromEntries(Object.keys(roleKinds).map(role => [role, nodeForRole(spec, role)]));
    const requiredRoutes = [
        ['execute', 'pass', 'verify'], ['execute', 'fail', 'execute'], ['execute', 'exhausted', 'human_gate'],
        ['verify', 'retry', 'execute'], ['verify', 'approve', 'human_gate'], ['verify', 'auto_promote', 'promote'], ['verify', 'exhausted', 'human_gate'],
        ['human_gate', 'approve', 'promote'], ['human_gate', 'retry', 'execute'], ['human_gate', 'reject', 'failed'], ['promote', 'pass', 'complete'],
    ];
    for (const [sourceRole, outcome, targetRole] of requiredRoutes)
        if (nextNode(spec, roles[sourceRole], outcome, 0) !== roles[targetRole])
            throw new Error(`LoopSpec route ${sourceRole}/${outcome} must target ${targetRole}`);
    const verifierFail = nextNode(spec, roles.verify, 'fail', 0);
    if (![roles.execute, roles.human_gate].includes(verifierFail))
        throw new Error('LoopSpec verifier/fail must target execute or human_gate');
}
